### PR TITLE
Adds in richer configuration of UI Table columns

### LINKED ISSUE #392: ui-table column type selection
### Description

Dash v1 ui-table allows selection of 'Link' for URL strings and 'HTML' for displaying HTML formatted strings.
Dash v2 ui-table does not offer any column types thus none of the URLs or HTML data is displayed correctly.

Request v2 ui-table widget be aligned with similar features as v1 widget.

### Properties

_No response_

### Events

_No response_

### Controls

_No response_

### Existing Examples

Example of ui-table column selection:
image

Displayed in v1 dash as follows:
image
 
### Have you provided an initial effort estimate for this issue?

I am no FlowFuse team member

### PR DESCRIPTION
**Pull request by joepavitt**
## Description

- Adds new configuration options on a column:
    - Width
    - Align
    - Type
- Updated the UI Data Table vue component to support the following cell types:
    - Text
    - HTML
    - Link
    - Progress
    - Color
    - Tick/Cross
    - Sparkline - Trend
    - Sparkline - Bar
    - Row Number
- Adds in-Editor documentation for UI Table
- Updates the online documentation for UI Table, including examples

### Example

Screenshot 2024-05-17 at 18 27 01

## Related Issue(s)

Closes #392

### CODE REVIEW

**Comment by thebaldgeek** [comment]
Humble and grateful thanks for this update Joe.
With the search option you added the past week and some responsiveness this ui-table node might just be the one stop node for 95% of all table needs in a dashboard.

**Comment by Steve-Mcl** [comment]
Overall, some nifty niceness, but I do have handful of suggestions/FRs (this or future iteration obvs ;) )

Upfront I just want to say sorry if anything comes across as negative (I am really trying to be constructive), I just want more sir.  So sorry, sorry, sorry (thats for the 3 FRs ;) ) 

###  UL LI  Layout
Element sizing in the UL LI entries is somewhat unattractive
chrome_wdLmCW1ZPc

For reference, here is how i laid out Cron Plus
chrome_wKuhHm4x24

And Buffer Parser: Note the use of `typedInputs` to avoid having separate labels
chrome_uxQFE53O1Z

<br>
<br>

### Feature Request
I truly hate having to add a change node before UI nodes and having to squirrel away my original data to an alt property of the message, then  copy the data from `payload.data` to `payload` blah blah. 
I find it unintuitive, unfriendly and not pursuant to low-code. There are LOTS of beginner questions on the forum where they go down some convoluted solution of storing the data in context, or branching wires or splitting etc to get around this short coming.
I would appreciate being able to specify where the table data is in the `msg` - in this example, `msg.payload.data`


### Feature Request
Being able to use sub properties as a key so that i dont have to mangle my data before the UI node

#### My data  - I want to use `costs.total` in the table

image

Specifying paths as the key (e.g. `costs.total`) would IMO, greatly improve usability / user friendliness (we could take advantage of [`RED.util.getObjectProperty`](https://nodered.org/docs/api/modules/v/0.20.0/@node-red_util_util.html#.getObjectProperty) to do this for us)

New users really struggle with data reformatting to suit the next node (ref: chart nodes etc)

### Stretch Feature Request

Support for data formatting would be a very nice touch. E.g. if i send a percentage number 0.5, it is shown as `0.5`, With a percentage formatter it would show "50%".  I realise Vue 3 dropped "filters" which would have been nice, but perhaps the "text" type could have an accompanying "format" dropdown containing things like "none", "percent", "yyyy/mm/dd", and various other common text formatting?  Just enough to ease the burden on the flow developer.

**Comment by joepavitt** [comment]
## Layout Feedback

Thanks for spotting that, which browser are you using (I'm assuming Windows?) It's quite different to what I'm seeing, but will do my best to resolve form my own machine.

I like the full width table/configuration list, so will copy that structure, and then tidy up the responsiveness of the data fields themselves. I'm not keen on the idea of using a TypedInput for this given that there is no need for the functionality they provide here.

## Feature Requests 

For all of the feature request - here isn't an appropriate place to detail them - they belong in their own issues. I will give a quick overview on each though:

- **msg.<property>:** Not a trivial change by any means, mostly due to how tied we are with `msg.payload` on client-side for many many widgets
- **Nested property:** Yes, this is definitely something we should support
- **Data Formatting:** It's been requested a lot, valid feature request, but not a feature that is required for this PR to merge though. It should be captured with a generalised architecture for how on earth we're going to solve that across the app for other properties too.

**Comment by joepavitt** [comment]
I have a min-width being set by some JS in Node-RED that prevents me from going as narrow as you're able to (I did speak with @knolleary about this on Friday) - if I override it, with some incoming changes I'll push momentarily, the tray looks as follows when force-narrowed `150px` beyond `min-width`

Screenshot 2024-05-19 at 15 30 20

**Comment by joepavitt** [comment]
@Steve-Mcl opened https://github.com/FlowFuse/node-red-dashboard/pull/875 which implements the nested property feature. Did as a standalone PR to make the reviews easier. Merge order doesn't matter as can update the other one to point to `main` if this merges first

**Comment by thebaldgeek** [comment]
@Steve-Mcl I'm stunned that you figured out how to test this node before it was merged. Well done. Looking forward to trying it out myself.

In regards to your msg.payload comments, I can not follow them. I just don't understand what you mean about msg.payload.data.
The only reason I struggled with dash 1.0 table node was because I was trying to use it before I understood what a JSON array was.
Once I got that sorted https://discourse.nodered.org/t/example-flow-dashboard-table-node/27061 it went a lot smoother / quicker and with less code.
If you want me / everyone to use a function node before the table node to change my msg.payload array to something that is the same as you use, then no problem. It just needs to be clear.
Terry and I wrote this video script to help me understand how to use the table node: https://discourse.nodered.org/t/intro-to-json-for-node-red-video/59997

I make heavy use of the FIFO node to build every table on the website. Simply inject one payload array and set the size of the table in the FIFO subflow node and its done. https://flows.nodered.org/flow/d7eea525606ea45a7c850976a85b349e
Its so simple and so low-code its nuts.
image
As you can see, its a clean one to one mapping from array to table.
It might just be a programmer vs low-code thinking.... Sort of a flat vs nested thing.....

I think if the node help and example library flows for the node shows your required array construction, it should be easy for everyone to use as we are all used to making custom tweaks to our payload since most of the more advanced nodes seem to require something slightly different.

Looking forward to testing the node and find the conversation interesting in describing how we all get things done.

**Comment by joepavitt** [comment]
> I'm stunned that you figured out how to test this node before it was merged

You can follow local install guide here: https://dashboard.flowfuse.com/contributing/#installing-locally if you're interested @thebaldgeek 

Anyone can install it locally, independent from the Node-RED Palette Manager - mostly useful for local development and Pull Request reviews, as I had asked Steve to do here.

**Comment by Steve-Mcl** [comment]
> which browser are you using (I'm assuming Windows?)

Chrome. And yes, on windows.  I'm my experience chrome on windows and Linux render the same. Not sure if some forces their renderer on Mac like they do on iDevices so not sure how you don't see it. Fwiw, that was running in v4 beta 3



> * **msg.:** Not a trivial change by any means, mostly due to how tied we are with `msg.payload` on client-side for many many widgets

I don't think it's non trivial tbh. Just let the user decide. It can be defaulted to msg.payload.


Regarding layout, I didn't call it out but hoped you'd spot it. The demos I posted size the ul container to the bottom of the form. It keeps the `add` button in view.

**Comment by Steve-Mcl** [comment]
> In regards to your msg.payload comments, I can not follow them. I just don't understand what you mean about msg.payload.data.
> The only reason I struggled with dash 1.0 table node was because I was trying to use it before I understood what a JSON array was.
> Once I got that sorted https://discourse.nodered.org/t/example-flow-dashboard-table-node/27061 it went a lot smoother / quicker and with less code.
> If you want me / everyone to use a function node before the table node to change my msg.payload array to something that is the same as you use, then no problem. It just needs to be clear

@thebaldgeek 

Some times, the data you want to display is NOT in msg.payload. Sometimes it is in msg.payload.something.else (i.e a sub property) or it is in msg.data.  

The request I am making lets you simply tell the ui-table node where the data is. 

Without this feature, a user HAS to mess about with a change node or function node BEFORE the ui-table.

**Comment by joepavitt** [comment]
> Regarding layout, I didn't call it out but hoped you'd spot it. The demos I posted size the ul container to the bottom of the form. It keeps the add button in view.

I did spot that, but having to manage/pin scroll to bottom each time you click add, and having only a small windows for the columns (which are each 4 rows of data) I chose to keep as-is for now

**Comment by joepavitt** [comment]
@Steve-Mcl fyi - fix is in for checkboxes, tests are passing

File: docs/examples/ui-table-cell-types.json [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            },
+            {
+                "title": "URL",
+                "key": "link",
+                "type": "link",
+                "width": "300px",
+                "align": "start"
+            },
+            {
+                "title": "Bar",
```
Comment (Steve-Mcl) [inline docs/examples/ui-table-cell-types.json:39]:
```suggestion
                "title": "Progress",
```

File: nodes/widgets/locales/en-US/ui_table.html [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<script type="text/html" data-help-name="ui-table">
+    <p>
+        Display and manage complex datasets with ease using the ui-table widget.
+    </p>
```
Comment (Steve-Mcl) [inline nodes/widgets/locales/en-US/ui_table.html:4]: Ideally, help would explain what the input data shape should be?

**Review by Steve-Mcl — APPROVED** [review]
Approved but with suggestions

* The demo has title "Bar" for "Progress". If you correct this, the image should also be updated
* The built-in help doesnt mention the input data should be an array of objects

Optional
* The input data is not sanity checked and so cryptic errors are shown in the console. Catching this (or pre-checking the input data is valid) and outputting a warning would be a bit better for the inevitable "my table doesnt work" forum questions

File: nodes/widgets/locales/en-US/ui_table.html (hunk 2 above)
Comment (joepavitt) [reply nodes/widgets/locales/en-US/ui_table.html:4]: Resolved in https://github.com/FlowFuse/node-red-dashboard/pull/873/commits/df1af51d3d29e3cabfbac018a9d85f9721a45afa

**Pull request merged**